### PR TITLE
Implement null model

### PR DESCRIPTION
**Pull request by aofarrel**
Includes checker workflow.

Known quirks:
* Does not include the phenotype metadata inheritance function, as it appears it would not work off SB and does not appear to make any difference to the output phenotype file. This is documented in the code itself and where appropriate in this repo.
* The output null model file tends to vary from what you'd get on SB. It seems its in the realm of a rounding error but we have only looked at one such example in detail.

### CODE REVIEW

**Comment by aofarrel** [comment]
Merged PR in which the checker model is now powered by R.

**Review by juklucas — CHANGES_REQUESTED** [review]
Looks great. At this point, I'd make changes similar to what I suggested for n_pcs for:
- family
- inverse_normal
- rescale_variance
- n_pcs
- resid_covars

At that point Avani and I can restart the review process.

File: null-model/null-model-wf.wdl [hunk 1]
```diff
@@ -0,0 +1,497 @@
...
+		
+		if "~{isdefined_group}" == "true":
+			f.write('group_var "~{group_var}"\n')
+		
+		if "~{isdefined_inverse}" == "true":
+			f.write('inverse_normal ~{inverse_normal}\n')
+		
+		if "~{isdefined_npcs}" == "true":
+			if int("~{n_pcs}") > 0:  # must be done this way or else syntax err when n_pcs not defined
+				f.write('n_pcs ~{n_pcs}\n')
```
[the remark this answers is not in the document]
Comment (aofarrel) [reply null-model/null-model-wf.wdl:150]: But it wouldn't match what Stephanie does. Stephanie does not check for a default value, she instead checks for whether or not something is defined. It's a minor distinction but it's actually behind a couple of bugs in the CWL so I think it's a distinction we need to keep making.

File: null-model/null-model-wf.wdl [hunk 2]
```diff
@@ -0,0 +1,497 @@
...
+	#
+	# Strictly speaking this is only needed for Array variables
+	# But we'll do it for most of 'em for consistency's sake
+	Boolean isdefined_conditvar = defined(conditional_variant_file)
+	Boolean isdefined_covars = defined(covars)
+	Boolean isdefined_gds = defined(gds_files)
+	Boolean isdefined_group = defined(group_var)
+	Boolean isdefined_inverse = defined(inverse_normal)
+	Boolean isdefined_matrix = defined(relatedness_matrix_file)
+	Boolean isdefined_npcs = defined(n_pcs)
```
[the remark this answers is not in the document]
Comment (aofarrel) [reply null-model/null-model-wf.wdl:259]: https://github.com/DataBiosphere/analysis_pipeline_WDL/pull/29#discussion_r668926479

File: null-model/null-model-wf.wdl [hunk 3]
```diff
@@ -0,0 +1,497 @@
...
+		File phenotype_file
+		String family  # required on SB but not in original pipeline
+
+		# optional stuff
+		File? conditional_variant_file
+		Array[String]? covars
+		Array[File]? gds_files
+		String? group_var
+		Boolean? inverse_normal
+		Int? n_pcs
```
[the remark this answers is not in the document]
Comment (aofarrel) [reply null-model/null-model-wf.wdl:34]: https://github.com/DataBiosphere/analysis_pipeline_WDL/pull/29#discussion_r668926479

File: null-model/README.md [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+* Null model output file
+
+## Common Issues
+* If **PCA File** is not provided, the **Number of PCs to include as covariates** parameter **must** be set to 0.
+* **PCA File** must be an RData object output from the *pcair* function in the GENESIS package.
+* The null model job can be very computationally demanding in large samples (e.g. > 20K). GENESIS supports using sparse representations of matrices in the **relatedness matrix file** via the R Matrix package, and this can substantially reduce memory usage and CPU time.
+* Keep in mind that **relatedness matrix file** can be a gds in the GRM case, but is usually an RData matrix.
+
+
+## More Information
```
[the remark this answers is not in the document]
Comment (aofarrel) [reply null-model/README.md:112]: Thank you!

File: null-model/checker/are_outputs_kinda_equal.R [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+options <- commandArgs(trailingOnly = TRUE)
+print(options)
+
+test <- loadRData(options[1])
+truth <- loadRData(options[2])
+toler <- as.numeric(options[3])
+
+# check to make sure we didn't accidentally load the same file twice
+# files should not be equivalent, as they failed an MD5
+stopifnot(isFALSE(identical(test, truth)))
```
[the remark this answers is not in the document]
Comment (aofarrel) [reply null-model/checker/are_outputs_kinda_equal.R:17]: This line is very useful in debugging situations where there is more than one output file, which in previous versions of this code was very error-prone. I've ironed out all the bugs I could find, meaning it is no longer required, but I'm hesitant to remove it considering I want to use this code for more complicated checker workflows later down the line.

File: _test-data-and-truths_/null_model/README.md [hunk 6]
```diff
@@ -0,0 +1 @@
+These files were taken straight from the original Python pipeline's repo, with the exception of the two Null_model_mixed* files which came from the GENESIS Null Model sample workspace, based upon Revision 17 of the CWL null model workflow, on Seven Bridges. The details of their creation has been lost to time, but it is known they are based on open data. It is not recommend to use these for your "real" data; they are presented here as truth files.
```
[the remark this answers is not in the document]
Comment (aofarrel) [reply _test-data-and-truths_/null_model/README.md:1]: The files have been unchanged for quite some time, so I don't think pointing to a commit is worth it. But I'm still open to the idea. Take a look at the new readmes and see what you think.

File: null-model/null-model-wf.wdl [hunk 7]
```diff
@@ -0,0 +1,497 @@
+version 1.0
+
+# Most pipelines so far use a .config file to point requiste inputs to the Rscript.
```
[the remark this answers is not in the document]
Comment (aofarrel) [reply null-model/null-model-wf.wdl:3]: Ugh!

File: null-model/null-model-wf.wdl (hunk 1 above)
[the remark this answers is not in the document]
Comment (juklucas) [reply null-model/null-model-wf.wdl:156]: You are quite right, she does a check for whether or not the user has defined n_pcs. Considering that we assign a default value in the inputs section, this check doesn't really mean anything for us.

[the remark this answers is not in the document]
Comment (aofarrel) [reply null-model/null-model-wf.wdl:156]: I'm confused. We don't assign a default value in the inputs section.

[the remark this answers is not in the document]
Comment (aofarrel) [reply null-model/null-model-wf.wdl:156]: Solved via Slack -- sbg:toolDefaultValue doesn't seem to write anything, it just seems to show a UI element.

File: null-model/null-model-wf.wdl [hunk 8]
```diff
@@ -0,0 +1,497 @@
...
+		Rscript /usr/local/analysis_pipeline/R/null_model.R null_model.config
+	>>>
+
+	# Estimate disk size required -- recall most inputs are duplicated
+	Int phenotype_size = 2*ceil(size(phenotype_file, "GB"))
+	# todo: other files, etc
+	Int finalDiskSize = phenotype_size + addldisk
+
+	# defined workaround
+	#
```
[the remark this answers is not in the document]
Comment (aofarrel) [reply null-model/null-model-wf.wdl:186]: Accidentally put this in, then reverted it -- let me know if anything else gets lost.

**Pull request closed without merging**